[Checkpoint] Separate implementation into generator

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105207
* #105206
* #105205
* #105180
* #105176
* __->__ #105101

Separates the non-reentrant AC implementation into a generator so that
other APIs such as composable checkpoint API can use the generator as pre and
post forward logic.

Differential Revision: [D47419387](https://our.internmc.facebook.com/intern/diff/D47419387/)

rohan-varma committed:
[Checkpoint] Separate implementation into generator

Separates the non-reentrant AC implementation into a generator so that
other APIs such as composable checkpoint API can use the generator as pre and
post forward logic.

Differential Revision: [D47419387](https://our.internmc.facebook.com/intern/diff/D47419387/)

[ghstack-poisoned]

rohan-varma commented:
@albanD was wondering your thoughts on this and in particular errors during the fwd pass corrupting state as you mentioned in https://github.com/pytorch/pytorch/pull/101650. It seems that this implementation will be no different than the current impl of non-reentrant AC, in that errors during forward will simply raise and we won't reset any of the checkpoint state?

soulitzer commented on torch/utils/checkpoint.py:
```diff
@@ -447,15 +447,19 @@ def checkpoint(
+            next(gen)
+        except StopIteration:
+            return ret
+        except BaseException:
```
Why do we need this extra except?

rohan-varma replied:
yep this is not needed.

soulitzer reviewed (dismissed):
Pretty cool! As long as we don't keep the generator alive unnecessarily the ctx clean up should run.

soulitzer commented:
> As long as we don't keep the generator alive unnecessarily the ctx clean up should run.

It sounds like when this is used, we'd need to store this on the `checkpoint.state`. Is it a guarantee that in the case where we have an exception that is caught, we no longer hold any references to the generator there?

rohan-varma committed:
Update on "[Checkpoint] Separate implementation into generator"


Separates the non-reentrant AC implementation into a generator so that
other APIs such as composable checkpoint API can use the generator as pre and
post forward logic.

Differential Revision: [D47419387](https://our.internmc.facebook.com/intern/diff/D47419387/)

[ghstack-poisoned]

rohan-varma commented:
@soulitzer I was able to use `always_call` flag in the post hook to ensure we clear out the generator regardless of exception in the composable AC in the next diff. Would be great to get your thoughts on that

soulitzer reviewed (approved):
Sounds good!

rohan-varma committed:
Update on "[Checkpoint] Separate implementation into generator"


Separates the non-reentrant AC implementation into a generator so that
other APIs such as composable checkpoint API can use the generator as pre and
post forward logic.

Differential Revision: [D47419387](https://our.internmc.facebook.com/intern/diff/D47419387/)

[ghstack-poisoned]

rohan-varma commented:
@pytorchbot merge -f "CI done"
